dali/pipeline/operators/color/hsv.h

```diff
@@ -0,0 +1,217 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATORS_COLOR_HSV_H_
+#define DALI_PIPELINE_OPERATORS_COLOR_HSV_H_
+
+#include <vector>
+#include <memory>
+#include <string>
+#include "dali/core/static_switch.h"
+#include "dali/core/geom/mat.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/pipeline/operators/common.h"
+#include "dali/pipeline/operators/operator.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/kernels/tensor_shape_print.h"
+#include "dali/kernels/imgproc/pointwise/linear_transformation_cpu.h"
+
+
+namespace dali {
+namespace hsv {
+
+/**
+ * Names of arguments
+ */
+extern const std::string kHue;
+extern const std::string kSaturation;
+extern const std::string kValue;
+extern const std::string kOutputType;
+
+/**
+ * Color space conversion
+ */
+const mat3 Rgb2Yiq = {{
+                              {.299f, .587f, .114f},
+                              {.596f, -.274f, -.321f},
+                              {.211f, -.523f, .311f}
+                      }};
+
+// Inversion of Rgb2Yiq, but pre-calculated, cause mat<> doesn't do inversion.
+const mat3 Yiq2Rgb = {{
+                              {1, .956f, .621f},
+                              {1, -.272f, -.647f},
+                              {1, -1.107f, 1.705f}
+                      }};
+
+
+inline mat3 compose_hue(float hue /* hue hue hue */ ) {
+  const auto h_rad = hue * M_PI / 180;
+  mat3 ret = mat3::eye();  // rotation matrix
+  // Hue change in YIQ color space is a rotation along the Y axis
+  ret(1, 1) = cos(h_rad);
+  ret(2, 2) = cos(h_rad);
+  ret(1, 2) = -sin(h_rad);
+  ret(2, 1) = sin(h_rad);
+  return ret;
+}
+
+inline mat3 compose_saturation(float saturation) {
+  mat3 ret = mat3::eye();
+  // In the YIQ color space, saturation change is a
+  // uniform scaling in IQ dimensions
+  ret(1, 1) = saturation;
+  ret(2, 2) = saturation;
+  return ret;
+}
+
+
+inline mat3 compose_value(float value) {
+  // In the YIQ color space, value change is a
+  // uniform scaling across all dimensions
+  return mat3::diag(value);
+}
+
+}  // namespace hsv
+
+
+
+template <typename Backend>
+class Hsv : public Operator<Backend> {
+ public:
+  ~Hsv() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(Hsv);
+
+ protected:
+  explicit Hsv(const OpSpec &spec) :
+          Operator<Backend>(spec),
+          output_type_(spec.GetArgument<DALIDataType>(hsv::kOutputType)) {
+    GetSingleOrRepeatedArg(spec, hue_, hsv::kHue, batch_size_);
```

Comment:
This is bad. The arguments are not meant to be vectors, they should be either scalars or tensors.